### PR TITLE
Add pause/resume functionality to agent-sdk conversation system

### PR DESCRIPTION
**Pull request by malhotra5**
## Summary

This PR adds pause/resume functionality to the agent-sdk conversation system, allowing CLI applications to interrupt agent execution between steps (not mid-LLM completion) from another thread.

## Key Features

- **Thread-safe pause control**: Added `agent_paused` boolean flag to `ConversationState` 
- **Event tracking**: Created `PauseEvent` class to record pause requests in conversation history
- **Simple API**: Added `pause()` method to `Conversation` class that can be called from any thread
- **Graceful interruption**: Updated `run()` loop to check pause flag between agent steps
- **Resume capability**: Calling `run()` again after pause resumes execution from where it left off

### Core Changes

1. **ConversationState** (`openhands/sdk/conversation/state.py`):
   - Added `agent_paused: bool = False` flag for thread-safe pause control

2. **PauseEvent** (`openhands/sdk/event/llm_convertible.py`):
   - New event class extending `EventBase` (not LLM convertible as requested)
   - Records pause requests in conversation history with "user" source

3. **Conversation.pause()** (`openhands/sdk/conversation/conversation.py`):
   - Thread-safe method using ConversationState context manager
   - Sets pause flag and emits PauseEvent for tracking
   - Can be called from any thread (e.g., CLI main thread)

4. **Updated run() loop**:
   - Checks pause flag at start of each iteration
   - Breaks gracefully when paused, allowing resume via subsequent `run()` calls
   - Resets pause flag when paused to enable resume

### Usage Pattern for CLI

```python
# Start agent in background thread
def run_agent():
    conversation.run()

agent_thread = threading.Thread(target=run_agent, daemon=True)
agent_thread.start()

# Pause from main thread when user types 'pause'
conversation.pause()

# Resume by calling run() again
conversation.run()
```

## Testing

- **8 comprehensive tests** covering basic functionality, threading, confirmation mode interaction, and edge cases
- **All existing tests pass** (17/17) ensuring no regressions
- **Thread safety verified** with concurrent pause/run operations

## Backward Compatibility

- ✅ No breaking changes to existing APIs
- ✅ Maintains compatibility with confirmation mode patterns  
- ✅ Follows existing code patterns for thread safety
- ✅ All existing tests continue to pass

## Design Decisions

1. **PauseEvent as EventBase**: Following user request, PauseEvent extends EventBase (not LLMConvertibleEvent) since pause requests shouldn't be sent to LLM
2. **Thread-safe state management**: Uses existing ConversationState context manager pattern for consistency
3. **Pause between steps**: Interrupts between agent steps rather than mid-LLM completion, as requested
4. **Simple API**: Single `pause()` method that can be called from any thread

## Related Issues

Addresses the need for pause/resume functionality in CLI applications where the caller runs the agent in a background thread and needs to interrupt execution based on user input.

## Testing Instructions

```bash
# Run pause functionality tests
uv run pytest tests/sdk/conversation/test_pause_functionality.py -v

# Run all conversation tests to verify no regressions  
uv run pytest tests/sdk/conversation/ -v
```

@malhotra5 can click here to [continue refining the PR](https://app.all-hands.dev/conversations/4c9a15d6770d45289308aa530512fda4)

### CODE REVIEW

File: openhands/sdk/event/llm_convertible.py [hunk 1]
```diff
@@ -230,3 +230,13 @@ def __str__(self) -> str:
             else self.error
         )
         return f"{base_str}\n  Error: {error_preview}"
+
+
+class PauseEvent(EventBase):
```
Comment (malhotra5) [inline openhands/sdk/event/llm_convertible.py]: @openhands lets actually make `PauseEvent` llm convertible

**Comment by malhotra5** [comment]
@openhands let's write a example script in `/examples folder` that will demonstrate the ability to pause a conversation 

Here is a sample script

```
import threading

# ...

agent = Agent(llm=llm, tools=tools)
conversation = Conversation(agent)
conversation.send_message("Say hello to 'world'")

thread = threading.Thread(target=conversation.run)
while conversation.state.status == ConversationStatus.RUNNING:
    conversation.send_message("keep going! you can do it!")
    time.sleep(1)
    
thread.join()
```

We want it so that a keyboard interrupt will pause the agent

Note that there are other code changes being made while you work, so make sure you pull the latest changes before you're ready to push

File: openhands/sdk/conversation/conversation.py [hunk 2]
```diff
@@ -120,6 +123,12 @@ def run(self) -> None:
             # and check will we be able to execute .send_message
             # BEFORE the .run loop finishes?
             with self.state:
+                # Check for pause request (can be called from another thread)
+                if self.state.agent_paused:
+                    self.state.agent_paused = False  # Reset for next run
+                    logger.info("Agent execution paused")
+                    break
+
```
Comment (malhotra5) [inline openhands/sdk/conversation/conversation.py]: Instead of breaking here, move the `self.state.agent_paused` conditional into the conditions for the while loop

also at the start of conversation run set `self.state.agent_paused` to False

Comment (malhotra5) [reply openhands/sdk/conversation/conversation.py]: @openhands address this comment

**Comment by malhotra5** [comment]
@openhands fix the unit tests in this PR

only modify the unit tests; rewrite majority of them, as the logic that it was testing originally has changed

some to consider 

1. multiple pause method calls successively only create one `PauseEvent`
2. calling conversation.pause() while conversation.run() is still running in a separate thread will pause the agent
3. calling conversation.run() on an already paused agent will resume it

**Comment by malhotra5** [comment]
@openhands fix the failing unit tests; the original logic has changed so only modify the test file until its passing

**Comment by malhotra5** [comment]
@openhands just fix the failing lint issues

**Comment by malhotra5** [comment]
@openhands use the $GITHUB_TOKEN to read the logs from the failing github CI

fix the linting issue; why is it throwing this type error when clearly it matches the base class. the other test files mock similarly and don't seem to be facing this issue

**Comment by malhotra5** [comment]
@openhands read the diff of this pr

after the conversation has been paused make it so that `/resume` will restart the `conversation.run` loop

File: openhands/sdk/conversation/conversation.py [hunk 3]
```diff
@@ -135,6 +148,10 @@ def run(self) -> None:
             if iteration >= self.max_iteration_per_run:
                 break
 
+            with self.state:
+                if self.state.agent_paused:
+                    break
```
Comment (xingyaoww) [inline openhands/sdk/conversation/conversation.py]: Why do we need to add this line?

**Review by xingyaoww — APPROVED** [review]
LGTM! Just one question!

File: openhands/sdk/conversation/conversation.py [hunk 4]
```diff
@@ -135,6 +148,10 @@ def run(self) -> None:
...
+
```
Comment (xingyaoww) [inline openhands/sdk/conversation/conversation.py]: Doesn't seems necessary to me, since it will be break in the next loop anyway:
```suggestion
```

Image
